app/models/resource.rb

```diff
@@ -0,0 +1,128 @@
+      yield_cached_raster(processed_raster_opts) { |raster_file| yield raster_file }
+    else
+      yield_uncached_raster(processed_raster_opts) { |raster_file| yield raster_file }
+    end
+  end
+
+  # Note: In the future, this method is unlikely to be called by Triclops because
+  # Hyacinth or Derivativo will be sending or updating crop region data.
+  def extract_featured_region!
+    self.with_source_image_file do |source_image_file|
+      Imogen.with_image(source_image_file.path) do |img|
+        # We try to use at least 768 pixels from any image when generating a
+        # featured area crop so that we don't unintentionally get a tiny
+        # 10px x 10px crop that gets scaled up for users and looks blocky/blurry.
+        left_x, top_y, right_x, bottom_y = Imogen::Iiif::Region::Featured.get(img, 768)
+        x = left_x
+        y = top_y
+        w = right_x - left_x
+        h = bottom_y - top_y
+        self.update(featured_region: "#{x},#{y},#{w},#{h}")
+      end
+    end
+  end
+
+  # Pre-processes raster opts before they're used later in the conversion chain.
+  # Performs operations like converting 'featured' region into a specific crop
+  # region and aliasing 'color' quality as 'default' quality.
+  # @param raster_opts [Hash]
+  #   A hash of IIIF options (e.g. {identifer: '...', region: '...', size: '...', etc. }).
+  # @return [Hash] the processed version of raster_opts
+  def preprocess_raster_opts(raster_opts)
+    # duplicate processed_raster_opts so we don't modify the incoming argument
+    processed_raster_opts = raster_opts.dup
+
+    # {region: 'featured'} should be converted into {region: 'x,y,w,h'}
+    if processed_raster_opts[:region] == 'featured'
+      extract_featured_region! if self.featured_region.blank?
+      processed_raster_opts[:region] = self.featured_region
+    end
+
+    # {quality: 'default'} is an alias for {quality: 'color'}
+    processed_raster_opts[:quality] = 'color' if processed_raster_opts[:quality] == 'default'
+
+    processed_raster_opts
+  end
+
+  def yield_cached_raster(raster_opts)
+    # Get cache path
+    raster_file_path = Triclops::RasterCache.instance.cache_path(raster_opts)
+
+    unless File.exist?(raster_file_path)
+      # We use a blocking lock so that two processes don't try to to run the
+      # same file creation operation at the same time.
+      Triclops::Lock.instance.with_blocking_lock(raster_file_path) do
+        # We do a second File.exists? check inside of the lock to ensure
+        # that file generation does not occur if the file was created
+        # while waiting to establish this lock.
+        unless File.exist?(raster_file_path)
+          FileUtils.mkdir_p(File.dirname(raster_file_path))
+          self.with_source_image_file do |source_image_file|
+            Triclops::Raster.generate(source_image_file.path, raster_file_path, raster_opts)
+          end
+        end
+      end
+    end
+
+    yield File.new(raster_file_path)
```

Comment:
That's right.  The inner method yields a value and I need to yield that value in the outer method to continue to pass it along to the outer caller.  I wonder if I could do it differently if I passed a block instead of yielding.  Would you or @barmintor prefer that?